### PR TITLE
UserControllerBundleIT work

### PR DESCRIPTION
**Pull request by frankhuster**
All ITs now pass.
Still some work to be done to use Region & Subscription helpers.

### CODE REVIEW

File: region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java [hunk 1]
```diff
@@ -69,7 +69,12 @@ public String getSqlQuery() {
             }
         };
 
-        List<Language> languages = languageDataService.executeSQLQuery(queryExecution);
+        List<Language> languages;
+        try {
+            languages = languageDataService.executeSQLQuery(queryExecution);
```
Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java]: Don't forget to pull this try catch out.

File: region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java [hunk 2]
```diff
@@ -57,7 +57,7 @@ public String getSqlQuery() {
                         "from nms_languages l " +
                         "join nms_districts d on  d.language_id_oid = l.id " +
                         "join nms_states s on d.state_id_oid = s.id " +
-                        "join FLW_MODULE_JOIN_CIRCLES_STATES cxs on s.id = cxs.state_id and cxs.circle_id = ?";
+                        "join state_circles cxs on s.id = cxs.state_id and cxs.circle_id = ?";
```
Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java]: Can you change the table name to nms_circles_join_states

File: testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java [hunk 3]
```diff
@@ -453,50 +346,28 @@ private void createFlwWithLanguageLocationCodeNotInWhitelist() {
...
-        State state2 = new State();
-        state2.setName("State 2");
-        state2.setCode(2L);
-        state2.getDistricts().add(district2);
+        // Calling these will make sure the districts exist and will map the districts' language to their state
+        rh.newDelhiDistrict();
+        rh.mysuruDistrict();
 
-        stateDataService.create(state2);
+        circleDataService.doInTransaction(new TransactionCallbackWithoutResult() {
```
Comment (larubbio) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: I'm concerned that we have to do this since we aren't doing this in the non-test code.

File: testing/src/test/java/org/motechproject/nms/testing/it/utils/RegionHelper.java [hunk 4]
```diff
@@ -95,6 +96,39 @@ public District newDelhiDistrict() {
...
+        return district;
+    }
+
+
+    public District bangaloreDistrict() {
+        District bangalore = districtDataService.findById(4L);
+        if (bangalore == null) {
+            bangalore = new District();
+            bangalore.setName("Bengaluru");
+            bangalore.setRegionalName("Bengaluru");
```
Comment (larubbio) [inline testing/src/test/java/org/motechproject/nms/testing/it/utils/RegionHelper.java]: Should we add a name with non-latin characters just for kicks?

File: testing/src/test/resources/bootstrap.properties [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+########################################################################
+jms.queue.for.events=QueueForEvents
+jms.queue.for.scheduler=QueueForScheduler
+jms.broker.url=tcp://localhost:61616
+jms.maximumRedeliveries=0
+jms.redeliveryDelayInMillis=2000
+jms.concurrentConsumers=1
+jms.maxConcurrentConsumers=10
+jms.session.cache.size=10
+jms.cache.producers=false
```
Comment (larubbio) [inline testing/src/test/resources/bootstrap.properties]: Why was this file needed now, but not before?

Comment (frankhuster) [inline testing/src/test/resources/bootstrap.properties]: Koshal suggested that this was the cause for the sporadic total failure due to the framework not starting.

File: testing/src/test/java/org/motechproject/nms/testing/it/utils/RegionHelper.java (hunk 4 above)
Comment (frankhuster) [inline testing/src/test/java/org/motechproject/nms/testing/it/utils/RegionHelper.java]: Are there going to be any in the real life data?

File: testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java (hunk 3 above)
Comment (frankhuster) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: Yeah, and now the CircleServiceBundleIT seems to break when [testing a state with multiple circles](http://ci.motechproject.org/job/NMS/334/testReport/junit/org.motechproject.nms.testing.it.region/CircleServiceBundleIT/testMultipleCirclesSingleState/) grrr... We need to have that fixed at the platform level.

Comment (frankhuster) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: Should we consider using Set<> instead of List<>?

**Comment by frankhuster** [comment]
so much for bootstrap.properties fixing the framework not started error...

**Comment by frankhuster** [comment]
ok to test

File: testing/src/test/resources/bootstrap.properties (hunk 5 above)
Comment (koshalt) [inline testing/src/test/resources/bootstrap.properties]: I'm not sure how this was working before, but we should ideally be
providing this system level config for our tests to talk to the right
components. Some of our intermittent failures were related to the config
module not finding this file

**Comment by koshalt** [comment]
test this please

**Comment by frankhuster** [comment]
ok to test

**Comment by llavoie** [comment]
@koshalt darn, it looks like the fix you were experimenting with wasn't the right one? :(

**Comment by larubbio** [comment]
Should someone just pull out the table name portion of this and commit that?

**Comment by llavoie** [comment]
Is that the only urgent part of this PR?

**Comment by larubbio** [comment]
yes

File: region/src/main/java/org/motechproject/nms/region/domain/State.java [hunk 6]
```diff
@@ -35,6 +41,10 @@
     private List<District> districts;
 
     @Field
+    @Cascade(delete = true)
+    @Persistent(table = "nms_states_join_circles", mappedBy = "states", defaultFetchGroup = "true")
+    @Join(column = "state_id")
+    @Element(column = "circle_id")
     private List<Circle> circles;
```
Comment (llavoie) [inline region/src/main/java/org/motechproject/nms/region/domain/State.java]: @larubbio  Do we just need this change (lines 44-47) and the change in LanguageServiceImpl.java?

Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/domain/State.java]: and the query in LanguageServiceImpl that mentions the table name.

File: region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java [hunk 7]
```diff
@@ -57,7 +57,7 @@ public String getSqlQuery() {
                         "from nms_languages l " +
                         "join nms_districts d on  d.language_id_oid = l.id " +
                         "join nms_states s on d.state_id_oid = s.id " +
-                        "join FLW_MODULE_JOIN_CIRCLES_STATES cxs on s.id = cxs.state_id and cxs.circle_id = ?";
+                        "join nms_states_join_circles sxc on s.id = sxc.state_id and sxc.circle_id = ?";
```
Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java]: @llavoie This one.

File: region/src/main/java/org/motechproject/nms/region/domain/State.java (hunk 6 above)
Comment (llavoie) [inline region/src/main/java/org/motechproject/nms/region/domain/State.java]: Great, I'll do it now.

File: region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java (hunk 7 above)
Comment (koshalt) [inline region/src/main/java/org/motechproject/nms/region/service/impl/LanguageServiceImpl.java]: The test container failures should be resolved now. Now, we are back to just a single test failure which I didn't get to debug before the game :)

File: region/src/main/java/org/motechproject/nms/region/domain/Circle.java [hunk 8]
```diff
@@ -16,7 +15,6 @@
     private String name;
 
     @Field
-    @Persistent(mappedBy = "circles", defaultFetchGroup = "true")
```
Comment (larubbio) [inline region/src/main/java/org/motechproject/nms/region/domain/Circle.java]: Is there a reason this was moved from Circle to State?

File: testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java [hunk 9]
```diff
@@ -453,50 +346,29 @@ private void createFlwWithLanguageLocationCodeNotInWhitelist() {
...
-        stateDataService.create(state2);
+        //TODO: remove this when https://applab.atlassian.net/browse/MOTECH-1679 is fixed
+        circleDataService.doInTransaction(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus transactionStatus) {
+                circle.getStates().add(rh.delhiState());
+                circle.getStates().add(rh.karnatakaState());
+                circleDataService.update(circle);
+            }
+        });
```
Comment (larubbio) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: @llavoie It may need this to create the many-to-many relationship in the database.  Although will @koshalt be able to fix this PR now?

Comment (koshalt) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: @larubbio @llavoie I was able to resolve the test container failures but hitting a new test failure here - http://ci.motechproject.org/job/NMS/369/testReport/org.motechproject.nms.testing.it.region/CircleServiceBundleIT/testMultipleCirclesSingleState/

Comment (larubbio) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: Try updating CircleBundleServiceIT so setupData runs in a transaction?

Comment (koshalt) [inline testing/src/test/java/org/motechproject/nms/testing/it/api/UserControllerBundleIT.java]: Trying that now..getting closer. Will post an update in a few minutes

**Comment by koshalt** [comment]
Merging this PR to try and unblock IT effort. Still needs investigtion.
